data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java

```diff
@@ -217,13 +218,37 @@ private <T> Record<Event> getRecord(ConsumerRecord<String, T> consumerRecord) {
                 value = objectMapper.readValue(jsonParser, Map.class);
             } else if (schema == MessageFormat.PLAINTEXT) {
                 value = (String)consumerRecord.value();
+                plainTextMode = true;
+            } else if (schema == MessageFormat.JSON) {
+                value = objectMapper.convertValue(value, Map.class);
             }
         } catch (Exception e){
             LOG.error("Failed to parse JSON or AVRO record", e);
         }
-        data.put(key, value);
-
+        if (!plainTextMode) {
+            if (!(value instanceof Map)) {
+                data.put(key, value);
+            } else {
+                Map<String, Object> valueMap = (Map<String, Object>)value;
+                if (kafkaKeyMode == KafkaKeyMode.INCLUDE_AS_FIELD) {
+                    valueMap.put("kafka_key", key);
+                }
+                data = valueMap;
+            }
+        } else {
+            if (Objects.isNull(key)) {
+                key = DEFAULT_KEY;
+            }
+            data.put(key, value);
+        }
         event = JacksonLog.builder().withData(data).build();
+        EventMetadata eventMetadata = event.getMetadata();
+        if (kafkaKeyMode == KafkaKeyMode.INCLUDE_AS_METADATA) {
+            eventMetadata.setAttribute("kafka_key", key);
+        }
+        eventMetadata.setAttribute("kafka_topic", topicName);
+        eventMetadata.setAttribute("kafka_partition", partition);
```

Comment:
is partition# useful info in eventMetadata. It can change anytime.

Reply:
It can't change for the event that is already received.